### ISSUE TITLE
Go Docker: Connection establishment fails

### ISSUE THREAD

**Issue by lubocode** [gekigek99/minecraft-server-hibernation#31]
* Make Server version into a Docker ENV variable and parse it in main.

* Currently script advertises vanilla server. Previously, when neither advertising version number nor vanilla status, modded MC servers worked without a problem. Make this either dynamic or remove it from info string.
  * This refers to e.g. forge or bukkit servers (with additional mods installed)

**Edit:** The two things above are partly implemented with the config.json from the latest commits on master.
The actual problem seems to be earlier during the connection establishment of the join process.

**Comment by lubocode** [comment]
@gekigek99 do you have a source for the the [server info packet](https://github.com/gekigek99/minecraft-vanilla-server-hibernation/wiki/Server-first-packet-response) or did you just try stuff out and can point me in the right direction for the server type?

**Comment by lubocode** [comment]
Also, startup message is kind of broken in the unRAID docker log.
![image](https://user-images.githubusercontent.com/24975209/93807029-5ad66180-fc4a-11ea-8e6b-1dd5a8e51c84.png)

**Comment by lubocode** [comment]
I am straight up unable to connect to the docker server after implementing the changes from master into the docker branch.
![image](https://user-images.githubusercontent.com/24975209/93808661-92dea400-fc4c-11ea-9951-0f3810a5b7d5.png)
![image](https://user-images.githubusercontent.com/24975209/93808838-d3d6b880-fc4c-11ea-85ff-15471295a525.png)
I am using 1.14.4, though something doesn't line up as Minecraft suggests the client to be out of date:
![image](https://user-images.githubusercontent.com/24975209/93808977-fbc61c00-fc4c-11ea-84e8-d78e2089004e.png)

@gekigek99 I guess that this is somehow correlated to the advertising of a vanilla server and it being a modded forge server, but do you have any other ideas what changes could have broken something for the docker environment?

**Comment by gekigek99** [comment]
> 
> 
> @gekigek99 do you have a source for the the [server info packet](https://github.com/gekigek99/minecraft-vanilla-server-hibernation/wiki/Server-first-packet-response) or did you just try stuff out and can point me in the right direction for the server type?

no i just did raw research as they say and wrote in the wiki what I found.

to get the correct server version you need to:
- run the script in debug mode
- launch the minecraft server
- request server info with the client

the script will capture 2 parameters that you need to replace in the script to make it work again

(next week i am planning on writing an update to make the process automatic (i will use an external config file that will be automatically updated when a new version is available))

**Comment by gekigek99** [comment]
first try to see if the ping works again then try to connect to the server (so we can see if these are related problems)

Right now i can't dedicate much time to the issue as i have exams but next week i should hopefully be able to dig into this

**Comment by gekigek99** [comment]
> 
> 
> Also, startup message is kind of broken in the unRAID docker log.
> ![image](https://user-images.githubusercontent.com/24975209/93807029-5ad66180-fc4a-11ea-8e6b-1dd5a8e51c84.png)

i rellay don't know why it would do that ahaha... kind of strange but for now i guess it's not the main problem

**Comment by gekigek99** [comment]
> ![image](https://user-images.githubusercontent.com/24975209/93808661-92dea400-fc4c-11ea-9951-0f3810a5b7d5.png)

does the server disconnects you when you ask for server info? or when you try to join?

**Comment by lubocode** [comment]
> does it work with a non-modded server?

This is the next step I have to test. I just thought that I might ask if you had any idea, before going on a journey to test a multitude of configurations^^
> Right now i can't dedicate much time to the issue as i have exams but next week i should hopefully be able to dig into this

No worries. Focus on your exams, the old version is still up and running. I created a development tag on Dockerhub for testing new versions :)

**Comment by lubocode** [comment]
> > ![image](https://user-images.githubusercontent.com/24975209/93808661-92dea400-fc4c-11ea-9951-0f3810a5b7d5.png)
> 
> does the server disconnects you when you ask for server info? or when you try to join?

When trying to join. Somewhere must be some miscommunication, as the old version correctly displays my username and starts up the server, while the new version goes into the routine for asking for the server info and not the join routine...

Maybe there is a problem within the docker environment. I will have to try running the script on my PC directly and then when that works transitioning into a Docker container again.

**Comment by gekigek99** [comment]
> > > ![image](https://user-images.githubusercontent.com/24975209/93808661-92dea400-fc4c-11ea-9951-0f3810a5b7d5.png)
> > 
> > does the server disconnects you when you ask for server info? or when you try to join?  

> Maybe there is a problem within the docker environment. I will have to try running the script on my PC directly and then when that works transitioning into a Docker container again.

It's quite strange because the new version, apart from the ping/"outdated" warning (because at mojang they updated the serverProtocol to a value higher than 751), it works ok for me (with vanilla server and not in the docker env)

Try to find the "serverProtocol" you need with that procedure and so we can confirm at least that the "require server info" part of the script is working correctly

**Comment by lubocode** [comment]
I am currently testing some different configurations and encountered a different bug in the main go script.
@gekigek99 should I open a new issue for this?
I am using 1.16.2 and with the current code and only changed debug to true.

![image](https://user-images.githubusercontent.com/24975209/94840266-73980180-0418-11eb-9d53-965194289051.png)

**Comment by gekigek99** [comment]
@lubocode yes i'll open it

**Comment by lubocode** [comment]
After testing a multitude of configurations and comparing the different version of the code I think there is no problem with the script itself. Maybe with the new server protocol does something different and that gets blocked by my PC.
I have tightened security with MS Defender so maybe that blocks part of the traffic, but to be sure I will need to retest on a different machine.
Running the master script in Windows and connecting through localhost works.

The script version from six months ago continues to work in the Docker with 1.14.4 and forge.
Testing that one with 1.16.2 and connecting with a different machine are the last two things I can currently test.

After that I will go deeper with capturing the server response, as you mentioned.

> > @gekigek99 do you have a source for the the [server info packet](https://github.com/gekigek99/minecraft-vanilla-server-hibernation/wiki/Server-first-packet-response) or did you just try stuff out and can point me in the right direction for the server type?
> 
> no i just did raw research as they say and wrote in the wiki what I found.
> 
> to get the correct server version you need to:
> 
> * set serveStatus to "online" and run the script in debug mode
> * launch manually the minecraft server
> * request server info with the client
> 
> the script will capture 2 parameters that you need to replace in the script to make it work again
> 
> (next week i am planning on writing an update to make the process automatic (i will use an external config file that will be automatically updated when a new version is available))

**Comment by lubocode** [comment]
No matter what version is set in the script, the connection still works and also with modded servers.
I do however get the same error as above.
![image](https://user-images.githubusercontent.com/24975209/94848106-c88d4500-0423-11eb-980b-1c41b7b975c7.png)

**Comment by lubocode** [comment]
Also, testing with WSL (2) is not good, as there seem to be some other problems occuring. Maybe due to the special environment mix.

**Comment by lubocode** [comment]
After testing just with my machine I can say that the advertised server version and protocol do not have an impact on being able to connect to the server if the server version is different or the server is modded.
Since that has also been implemented with the config.json I will rename this issue.

**Comment by gekigek99** [comment]
this happens always on windows right?

is it the same problem of #34 ?

**Comment by gekigek99** [comment]
something i can tell you for sure is that in the last 2 images serverVersion and serverProtocol are written in the wrong way:
they should be `... serverVersion: 1.14.4 serverProtocol: 498` without any characters (go-version 3.4)

(remember always to check the config.json for errors like `751}}` or `"1.14.4"`)

**Comment by lubocode** [comment]
> something i can tell you for sure is that in the last 2 images serverVersion and serverProtocol are written in the wrong way:
> they should be `... serverVersion: 1.14.4 serverProtocol: 498` without any characters (go-version 3.4)
> 
> (remember always to check the config.json for errors like `751}}` or `"1.14.4"`)

Config looks good. The numbers are written in the same way as in the original.
```
    "ServerVersion": "1.14.4",
    "ServerProtocol": "498"
```



> (I'm not an expert for docker)

No problem. Me neither, so I just work with what I know^^
I will try to get another machine or OS installation tomorrow, because I'm away for a few days after that, but since the old Docker version still works, I may postpone it to the end of next week.

**Comment by gekigek99** [comment]
ok no problem!

could issue 31 and 34 be related in some way (like the same problem from different point of view)?

**Comment by gekigek99** [comment]
Can we just make the point of the situation again? ;)
What is the problem? when does it happen?
how can I reproduce it? (what script, what docker, how to install the docker, what steps do i need to take to make it happen...)

Just so that in the next days i'll have a base from where i can help a little more ;)

**Comment by lubocode** [comment]
> could issue 31 and 34 be related in some way (like the same problem from different point of view)?

I don't know. For now it seems like this issue indicates a problem **before** starting the server while issue 34 indicates a problem **after** starting the server and either detecting the version or a player leaving the server.

**Comment by lubocode** [comment]
> Can we just make the point of the situation again? ;)
> What is the problem? when does it happen?
> how can I reproduce it? (what script, what docker, how to install the docker, what steps do i need to take to make it happen...)
> 
> Just so that in the next days I'll have a base from where I can help a little more ;)

Sure, I will create a summary tomorrow afternoon, when its not 5 minutes to midnight and I can concentrate a bit better^^

**Comment by gekigek99** [comment]
> I don't know. For now it seems like this issue indicates a problem **before** starting the server while issue 34 indicates a problem **after** starting the server and either detecting the version or a player leaving the server.

okok but then how can you reach the problem **after** starting the server if the problem **before** starting the server prevents you from connecting at all?

> Sure, I will create a summary tomorrow afternoon, when its not 5 minutes to midnight and I can concentrate a bit better^^

ok thanks take your time

**Comment by lubocode** [comment]
> okok but then how can you reach the problem **after** starting the server if the problem **before** starting the server prevents you from connecting at all?

Well, that is because issue #34 _seems_ Windows (10?) specific, while this issue #31 is happening with the Docker image I tried building from merge #29 .
On Windows, I can join the server without problems, while with the Docker there is not even a connection recognized by the script that triggers the server start. From a quick test from a different PC yesterday, it seems at least partially to be a problem with my PC, as the server started, though after the startup countdown trying to join the server timed out. I will have to add more logging, as `log.Fatal(err)` worked in some instances, but not others, so that I didn't get enough info yesterday.

As for what you might be able to test... Maybe take the script from the docker branch and run it on a native Linux distro like Ubuntu. If that works, try installing the development docker with `docker pull lubocode/minecraftserver-hibernate:development` and following the README from the [dockerhub page](https://hub.docker.com/r/lubocode/minecraftserver-hibernate).
While you won't be able to update that, you can also just build the docker yourself with the Dockerfile from the docker branch and test out code with that. Keep in mind, that you will have to copy the MC server.jar into the volume specified by yourself upon installing the Docker container.

If there are no problems for you, don't sweat it. I will come back to it next week and do some more testing after implementing more logging. You can then just focus on what you would like to do with the master script, or if there is nothing and you feel bored, maybe add some code comments, so that it will be easier to understand for newcomers^^
A wiki is nice and all, but adding comments where code is not directly understandable or where variables only make sense with background knowledge is worth a lot for seamless group work on shared code.

When I come back next week, we can also have a skype/discord/whatever session and talk to further our understanding of these problems without making more of a mess through communicating with issue comments 😅
I have your mail address, so I will just contact you when I am back, such that we can share contact info for the desired voice service.

**Comment by gekigek99** [comment]
okok i understand better, for now i think i'll focus on solving issue #34 and adding comments to go version (and i think that will keep me busy enough ahah)

> add some code comments, so that it will be easier to understand for newcomers^^
> A wiki is nice and all, but adding comments where code is not directly understandable or where variables only make sense with background knowledge is worth a lot for seamless group work on shared code.

yep you are right i opened issue #36 for that (the wiki is more like for me to remember things though ahah)

> When I come back next week, we can also have a skype/discord/whatever session and talk to further our understanding of these problems without making more of a mess through communicating with issue comments

yes i think it's a more appropriate way of doing things ahah

>  I have your mail address, so I will just contact you when I am back, such that we can share contact info for the desired voice service.

great!

**Comment by gekigek99** [comment]
so... i've looked fastly at possible differences and i don't see where the problem might be...

now i'll refer to this image:
![image](https://user-images.githubusercontent.com/53654579/95899516-fada6300-0d90-11eb-99e2-008dc375608f.png)

you try to join and it says you that you requested server info then it closes connection

this is what happens on the script:

![Screenshot_2020-10-13 gekigek99 minecraft-server-hibernation](https://user-images.githubusercontent.com/53654579/95900595-68d35a00-0d92-11eb-911b-127d17389e00.png)

(the green is what it's doing the red it's what it should do)

we need to understand why buffer[dataLen-1] == 0/1 and not 2 in docker so i would try adding an `fmt.println(buffer[:dataLen])` just after `dataLen, err := clientSocket.Read(buffer)`

**Comment by lubocode** [comment]
```
2020/10/14 15:36:48 *** from 172.17.224.1:25555 to 127.0.0.1:25565
[27 0 242 3 20 49 55 50 46 49 55 46 50 51 52 46 50 52 56 0 70 77 76 50 0 99 211 2 11 0 9 76 117 107 97 115 117 107 52 108]
2020/10/14 15:36:48 closing connection for: 172.17.224.1
```

**Comment by lubocode** [comment]
Spamming the server with requests and cancelling them leads to a server startup eventually.
**Connecting to the started server does in fact work.**

**Comment by gekigek99** [comment]
> Trying to join (server still down, connection fails):
> 
> ```
> 2020/10/14 15:36:48 *** from 172.17.224.1:25555 to 127.0.0.1:25565
> [27 0 242 3 20 49 55 50 46 49 55 46 50 51 52 46 50 52 56 0 70 77 76 50 0 99 211 2 11 0 9 76 117 107 97 115 117 107 52 108]
> 2020/10/14 15:36:48 closing connection for: 172.17.224.1
> ```

this is the problem i explained to you yesterday where golang tends to put packets together...
on python it should be like this:  
```
[27 0 242 3 20 49 55 50 46 49 55 46 50 51 52 46 50 52 56 0 70 77 76 50 0 99 211 *2*]  
[11 0 9 76 117 107 97 115 117 107 52 108]
```

(the highlighted 2 is the one that should trigger the start command)

(the second packet is containing the player name after the `[9]` (it's interesting to notice how the player name is sent with the format that is described in the wiki -> `[11 0 9]` where 9 is the lenght of the player name))

**Comment by lubocode** [comment]
It now works. See commit d556e64b9e63f559c5c8398205a05620d71d0c8b for the fix for connection establishment.
Now the server doesn't start up correctly, but I will open another issue for that. This ones has been abused enough already xD

**Comment by gekigek99** [comment]
@lubocode as a side note:  
in the commit on Apr 7, 2020 (2708b6cc4ce9f399fc3d05400ae2c87635b21ee9) the go version still had a (really bad...) function called `FindPattern()` used in this way:  

```
locationpattern, foundpattern := FindPattern(buffer[:datalenght], []byte{2, 11, 0, 9}) //for me is 2,11,0,9 don't know if it's a general rule
if buffer[datalenght-1] == 2 || foundpattern {
    // player is trying to connect
}
```

i removed it after discovering how the header magic works (11 0 9) because i understood it worked only for playerName length of 9...  

this is why  for you the old version was working and the newer ones did not: you were lucky enough that your playerName is the same length as mine and therefore 11 0 9 where the correct numbers to make everything work for your specific case, but would not have been functional for others.
